File: c/src/report.c
/* report.c — CSV and JSON duplicate-report export. See report.h. */

#define _DEFAULT_SOURCE

#include "report.h"

#include <stdlib.h>
#include <string.h>
#include <time.h>

#define REPORT_SCHEMA "twintidy.duplicate-report/v1"

const char *td_report_extension(td_report_format format) {
    return format == TD_REPORT_JSON ? ".json" : ".csv";
}

/*
 * Spreadsheet software evaluates a cell beginning with one of these as a
 * formula or command. Prefixing with an apostrophe keeps the text visible
 * while preventing evaluation.
 */
static int needs_formula_guard(const char *value) {
    if (value == NULL || *value == '\0') {
        return 0;
    }
    switch (*value) {
    case '=':
    case '+':
    case '-':
    case '@':
    case '\t':
    case '\r':
        return 1;
    default:
        return 0;
    }
}

static void write_csv_cell(FILE *out, const char *value) {
    if (value == NULL) {
        value = "";
    }
    fputc('"', out);
    if (needs_formula_guard(value)) {
        fputc('\'', out);
    }
    for (const char *cursor = value; *cursor != '\0'; cursor++) {
        if (*cursor == '"') {
            fputc('"', out); /* RFC 4180 doubling */
        }
        fputc(*cursor, out);
    }
    fputc('"', out);
}

static void write_json_string(FILE *out, const char *value) {
    if (value == NULL) {
        value = "";
    }
    fputc('"', out);
    for (const unsigned char *cursor = (const unsigned char *)value; *cursor != '\0'; cursor++) {
        switch (*cursor) {
        case '"':
            fputs("\\\"", out);
            break;
        case '\\':
            fputs("\\\\", out);
            break;
        case '\n':
            fputs("\\n", out);
            break;
        case '\r':
            fputs("\\r", out);
            break;
        case '\t':
            fputs("\\t", out);
            break;
        default:
            if (*cursor < 0x20) {
                fprintf(out, "\\u%04x", *cursor);
            } else {
                fputc((int)*cursor, out);
            }
        }
    }
    fputc('"', out);
}

static void format_timestamp(int64_t seconds, char *out, size_t size) {
    if (seconds <= 0) {
        out[0] = '\0';
        return;
    }
    time_t value = (time_t)seconds;
    struct tm parts;
#ifdef _WIN32
    if (gmtime_s(&parts, &value) != 0) {
        out[0] = '\0';
        return;
    }
#else
    if (gmtime_r(&value, &parts) == NULL) {
        out[0] = '\0';
        return;
    }
#endif
    if (strftime(out, size, "%Y-%m-%dT%H:%M:%SZ", &parts) == 0) {
        out[0] = '\0';
    }
}

static void generated_at(char *out, size_t size) {
    format_timestamp((int64_t)time(NULL), out, size);
}

static uint64_t group_reclaimable(const td_group *group) {
    if (group->count < 2) {
        return 0;
    }
    return group->size * (uint64_t)(group->count - 1);
}

static int write_csv(FILE *out, const char *folder, const td_result *result) {
    char stamp[64];
    generated_at(stamp, sizeof stamp);

    fputs("generatedAt,scanFolder,group,sha256,groupSize,groupReclaimableBytes,"
          "reportReclaimableBytes,path,fileSize,modifiedAt,category\r\n",
          out);

    int report_total_written = 0;
    for (size_t g = 0; g < result->count; g++) {
        const td_group *group = &result->groups[g];
        for (size_t f = 0; f < group->count; f++) {
            const td_file *file = &group->files[f];
            char modified[64];
            format_timestamp(file->modified_at, modified, sizeof modified);

            write_csv_cell(out, stamp);
            fputc(',', out);
            write_csv_cell(out, folder);
            fprintf(out, ",%zu,", g + 1);
            write_csv_cell(out, group->hash);
            fprintf(out, ",%llu,", (unsigned long long)group->size);
            /* Per-group and per-report totals are emitted once each, so a
             * reader can sum the file rows without double counting. */
            if (f == 0) {
                fprintf(out, "%llu", (unsigned long long)group_reclaimable(group));
            }
            fputc(',', out);
            if (!report_total_written) {
                fprintf(out, "%llu", (unsigned long long)result->reclaimable);
                report_total_written = 1;
            }
            fputc(',', out);
            write_csv_cell(out, file->path);
            fprintf(out, ",%llu,", (unsigned long long)file->size);
            write_csv_cell(out, modified);
            fputc(',', out);
            write_csv_cell(out, td_category_label(file->category));
            fputs("\r\n", out);
        }
    }
    return ferror(out) == 0;
}

static int write_json(FILE *out, const char *folder, const td_result *result) {
    char stamp[64];
    generated_at(stamp, sizeof stamp);

    size_t file_count = 0;
    for (size_t g = 0; g < result->count; g++) {
        file_count += result->groups[g].count;
    }

    fputs("{\n  \"schema\": ", out);
    write_json_string(out, REPORT_SCHEMA);
    fputs(",\n  \"generatedAt\": ", out);
    write_json_string(out, stamp);
    if (folder != NULL && *folder != '\0') {
        fputs(",\n  \"folder\": ", out);
        write_json_string(out, folder);
    }
    fprintf(out, ",\n  \"groupCount\": %zu", result->count);
    fprintf(out, ",\n  \"fileCount\": %zu", file_count);
    fprintf(out, ",\n  \"reclaimableBytes\": %llu",
            (unsigned long long)result->reclaimable);
    fputs(",\n  \"groups\": [", out);

    for (size_t g = 0; g < result->count; g++) {
        const td_group *group = &result->groups[g];
        fputs(g > 0 ? ",\n    {" : "\n    {", out);
        fprintf(out, "\n      \"size\": %llu", (unsigned long long)group->size);
        fputs(",\n      \"sha256\": ", out);
        write_json_string(out, group->hash);
        fputs(",\n      \"files\": [", out);
        for (size_t f = 0; f < group->count; f++) {
            const td_file *file = &group->files[f];
            char modified[64];
            format_timestamp(file->modified_at, modified, sizeof modified);

            fputs(f > 0 ? ",\n        {" : "\n        {", out);
            fputs("\"path\": ", out);
            write_json_string(out, file->path);
            fprintf(out, ", \"size\": %llu", (unsigned long long)file->size);
            fputs(", \"modifiedAt\": ", out);
            write_json_string(out, modified);
            fputs(", \"category\": ", out);
            write_json_string(out, td_category_label(file->category));
            fputc('}', out);
        }
        fputs(group->count > 0 ? "\n      ]" : "]", out);
        fputs("\n    }", out);
    }
    fputs(result->count > 0 ? "\n  ]\n}\n" : "]\n}\n", out);
    return ferror(out) == 0;
}

int td_report_write(FILE *out, td_report_format format, const char *folder,
                    const td_result *result) {
    if (out == NULL || result == NULL) {
        return 0;
    }
    if (format == TD_REPORT_JSON) {
        return write_json(out, folder, result);
    }
    return write_csv(out, folder, result);
}

int td_report_write_file(const char *path, td_report_format format,
                         const char *folder, const td_result *result) {
    if (path == NULL || result == NULL) {
        return 0;
    }
    size_t length = strlen(path) + 5;
    char *staging = malloc(length);
    if (staging == NULL) {
        return 0;
    }
    snprintf(staging, length, "%s.tmp", path);

    FILE *out = fopen(staging, "wb");
    if (out == NULL) {
        free(staging);
        return 0;
    }
    int ok = td_report_write(out, format, folder, result);
    if (fclose(out) != 0) {
        ok = 0;
    }
    if (ok) {
        remove(path); /* rename() will not replace on Windows */
        ok = rename(staging, path) == 0;
    }
    if (!ok) {
        remove(staging);
    }
    free(staging);
    return ok;
}

File: rust/src/report.rs
//! Duplicate-report export, matching the Go engine's formats.
//!
//! CSV cells that spreadsheet software would evaluate as formulas are
//! neutralized, so an exported report cannot execute anything when opened.

use std::fmt::Write as _;
use std::fs;
use std::io;
use std::path::Path;

use crate::{format_timestamp, ScanResult};

pub const SCHEMA: &str = "twintidy.duplicate-report/v1";

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Format {
    Csv,
    Json,
}

impl Format {
    pub fn extension(self) -> &'static str {
        match self {
            Format::Csv => ".csv",
            Format::Json => ".json",
        }
    }

    pub fn from_name(name: &str) -> Option<Format> {
        match name.trim().to_ascii_lowercase().as_str() {
            "csv" => Some(Format::Csv),
            "json" => Some(Format::Json),
            _ => None,
        }
    }
}

/// Spreadsheet software evaluates a cell beginning with one of these as a
/// formula or command. Prefixing with an apostrophe keeps the text visible
/// while preventing evaluation.
fn needs_formula_guard(value: &str) -> bool {
    matches!(
        value.chars().next(),
        Some('=') | Some('+') | Some('-') | Some('@') | Some('\t') | Some('\r')
    )
}

fn csv_cell(value: &str) -> String {
    let mut cell = String::with_capacity(value.len() + 2);
    cell.push('"');
    if needs_formula_guard(value) {
        cell.push('\'');
    }
    for character in value.chars() {
        if character == '"' {
            cell.push('"'); // RFC 4180 doubling
        }
        cell.push(character);
    }
    cell.push('"');
    cell
}

fn json_string(value: &str) -> String {
    let mut out = String::with_capacity(value.len() + 2);
    out.push('"');
    for character in value.chars() {
        match character {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if (c as u32) < 0x20 => {
                let _ = write!(out, "\\u{:04x}", c as u32);
            }
            c => out.push(c),
        }
    }
    out.push('"');
    out
}

/// Render the report for `result`. `folder` labels the scan root.
/// `generated_at` is a Unix timestamp so callers control the clock.
pub fn render(format: Format, folder: &str, result: &ScanResult, generated_at: i64) -> String {
    match format {
        Format::Csv => render_csv(folder, result, generated_at),
        Format::Json => render_json(folder, result, generated_at),
    }
}

fn render_csv(folder: &str, result: &ScanResult, generated_at: i64) -> String {
    let stamp = format_timestamp(generated_at);
    let mut out = String::new();
    out.push_str(
        "generatedAt,scanFolder,group,sha256,groupSize,groupReclaimableBytes,\
         reportReclaimableBytes,path,fileSize,modifiedAt,category\r\n",
    );

    let mut report_total_written = false;
    for (index, group) in result.groups.iter().enumerate() {
        let group_reclaimable = if group.files.len() >= 2 {
            group.size * (group.files.len() as u64 - 1)
        } else {
            0
        };
        for (position, file) in group.files.iter().enumerate() {
            // Per-group and per-report totals are emitted once each, so a
            // reader can sum the file rows without double counting.
            let group_cell = if position == 0 {
                group_reclaimable.to_string()
            } else {
                String::new()
            };
            let report_cell = if report_total_written {
                String::new()
            } else {
                report_total_written = true;
                result.reclaimable.to_string()
            };
            let _ = writeln!(
                out,
                "{},{},{},{},{},{},{},{},{},{},{}\r",
                csv_cell(&stamp),
                csv_cell(folder),
                index + 1,
                csv_cell(&group.hash),
                group.size,
                group_cell,
                report_cell,
                csv_cell(&file.path.to_string_lossy()),
                file.size,
                csv_cell(&format_timestamp(file.modified_at)),
                csv_cell(file.category.label()),
            );
        }
    }
    out
}

fn render_json(folder: &str, result: &ScanResult, generated_at: i64) -> String {
    let file_count: usize = result.groups.iter().map(|group| group.files.len()).sum();
    let mut out = String::new();
    let _ = write!(
        out,
        "{{\n  \"schema\": {},\n  \"generatedAt\": {}",
        json_string(SCHEMA),
        json_string(&format_timestamp(generated_at))
    );
    if !folder.is_empty() {
        let _ = write!(out, ",\n  \"folder\": {}", json_string(folder));
    }
    let _ = write!(
        out,
        ",\n  \"groupCount\": {},\n  \"fileCount\": {},\n  \"reclaimableBytes\": {},\n  \"groups\": [",
        result.groups.len(),
        file_count,
        result.reclaimable
    );

    for (index, group) in result.groups.iter().enumerate() {
        out.push_str(if index > 0 { ",\n    {" } else { "\n    {" });
        let _ = write!(
            out,
            "\n      \"size\": {},\n      \"sha256\": {},\n      \"files\": [",
            group.size,
            json_string(&group.hash)
        );
        for (position, file) in group.files.iter().enumerate() {
            out.push_str(if position > 0 {
                ",\n        {"
            } else {
                "\n        {"
            });
            let _ = write!(
                out,
                "\"path\": {}, \"size\": {}, \"modifiedAt\": {}, \"category\": {}}}",
                json_string(&file.path.to_string_lossy()),
                file.size,
                json_string(&format_timestamp(file.modified_at)),
                json_string(file.category.label())
            );
        }
        out.push_str(if group.files.is_empty() {
            "]"
        } else {
            "\n      ]"
        });
        out.push_str("\n    }");
    }
    out.push_str(if result.groups.is_empty() {
        "]\n}\n"
    } else {
        "\n  ]\n}\n"
    });
    out
}

/// Write atomically: the report is built beside `path` and renamed into
/// place only once complete, so a failure never leaves a truncated report
/// at the destination.
pub fn write_file(
    path: &Path,
    format: Format,
    folder: &str,
    result: &ScanResult,
    generated_at: i64,
) -> io::Result<()> {
    let staging = path.with_extension(format!(
        "{}tmp",
        path.extension()
            .map(|value| format!("{}.", value.to_string_lossy()))
            .unwrap_or_default()
    ));
    fs::write(&staging, render(format, folder, result, generated_at))?;
    match fs::rename(&staging, path) {
        Ok(()) => Ok(()),
        Err(error) => {
            let _ = fs::remove_file(&staging);
            Err(error)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{Category, DuplicateGroup, FileRecord};
    use std::path::PathBuf;

    fn sample() -> ScanResult {
        ScanResult {
            groups: vec![DuplicateGroup {
                size: 10,
                hash: "abc123".into(),
                files: vec![
                    FileRecord {
                        path: PathBuf::from("=danger.txt"),
                        size: 10,
                        modified_at: 1_784_937_600,
                        category: Category::Text,
                    },
                    FileRecord {
                        path: PathBuf::from("plain.txt"),
                        size: 10,
                        modified_at: 1_784_937_600,
                        category: Category::Text,
                    },
                ],
            }],
            reclaimable: 10,
            files_considered: 2,
            files_unreadable: 0,
        }
    }

    #[test]
    fn csv_neutralizes_formula_cells() {
        let csv = render(Format::Csv, "scan", &sample(), 1_784_937_600);
        assert!(
            csv.contains("\"'=danger.txt\""),
            "formula-leading path was not neutralized:\n{csv}"
        );
        // No cell may begin with a raw formula character.
        assert!(!csv.contains(",=") && !csv.contains("\n="));
    }

    #[test]
    fn csv_emits_totals_once() {
        let csv = render(Format::Csv, "scan", &sample(), 1_784_937_600);
        let rows: Vec<&str> = csv
            .lines()
            .skip(1)
            .filter(|line| !line.is_empty())
            .collect();
        assert_eq!(rows.len(), 2);
        // The group and report totals appear on the first row only.
        assert!(rows[0].contains(",10,10,10,"));
        assert!(rows[1].contains(",10,,,"));
    }

    #[test]
    fn json_is_well_formed_and_escaped() {
        let json = render(Format::Json, "scan", &sample(), 1_784_937_600);
        assert!(json.starts_with('{') && json.trim_end().ends_with('}'));
        assert!(json.contains("\"schema\": \"twintidy.duplicate-report/v1\""));
        assert!(json.contains("\"groupCount\": 1"));
        assert!(json.contains("\"fileCount\": 2"));
        // Backslashes in Windows paths must be escaped.
        let escaped = render(Format::Json, r"C:\scan", &sample(), 1_784_937_600);
        assert!(escaped.contains(r#""folder": "C:\\scan""#));
    }

    #[test]
    fn empty_result_still_renders_valid_documents() {
        let empty = ScanResult::default();
        let json = render(Format::Json, "", &empty, 1_784_937_600);
        assert!(json.contains("\"groups\": []"));
        let csv = render(Format::Csv, "", &empty, 1_784_937_600);
        assert_eq!(csv.lines().count(), 1, "header only");
    }

    #[test]
    fn format_names_round_trip() {
        assert_eq!(Format::from_name("JSON"), Some(Format::Json));
        assert_eq!(Format::from_name("csv"), Some(Format::Csv));
        assert_eq!(Format::from_name("xml"), None);
        assert_eq!(Format::Json.extension(), ".json");
    }
}
